host/xtest/asym_cipher_perf.c

```diff
@@ -0,0 +1,850 @@
+}
+
+static uint32_t get_curve_id(uint32_t width_bits)
+{
+	uint32_t curve_id = TEE_ERROR_BAD_PARAMETERS;
+
+	switch (width_bits) {
+	case ECC_CURVE_192:
+		curve_id = TEE_ECC_CURVE_NIST_P192;
+		break;
+	case ECC_CURVE_224:
+		curve_id = TEE_ECC_CURVE_NIST_P224;
+		break;
+	case ECC_CURVE_256:
+		curve_id = TEE_ECC_CURVE_NIST_P256;
+		break;
+	case ECC_CURVE_384:
+		curve_id = TEE_ECC_CURVE_NIST_P384;
+		break;
+	case ECC_CURVE_521:
+		curve_id = TEE_ECC_CURVE_NIST_P521;
+		break;
+	default:
+		printf("ECC curve is not support!\n");
+		break;
+	}
+
+	return curve_id;
+}
+
+static int asym_cipher_perf_run_test(int mode, size_t size, uint32_t n,
+				     uint32_t l, int is_random, uint32_t warmup,
+				     uint32_t verbosity, uint32_t width_bits,
+				     uint32_t tee_type, uint32_t salt_len,
+				     uint32_t algo)
+{
+	TEEC_Operation op = TEEC_OPERATION_INITIALIZER;
+	uint32_t cmd = TA_CRYPTO_PERF_CMD_ASYM_CIPHER_PROCESS;
+	TEEC_Result res = TEEC_ERROR_GENERIC;
+	uint8_t keygen_dh_p[DH_MAX_SIZE] = { };
+	uint8_t keygen_dh_g[DH_G_SIZE] = { };
+	TEE_Attribute params[4] = { };
+	size_t param_count = 0;
+	uint32_t curve_id = 0;
+	struct statistics stats = { };
+	struct timespec ts = { };
+	struct timespec t0 = { };
+	struct timespec t1 = { };
+	uint32_t ret_origin = 0;
+	uint8_t *buf = NULL;
+	size_t blen = 0;
+	double sd = 0;
+	int n0 = n;
+
+	if (clock_getres(CLOCK_MONOTONIC, &ts) < 0) {
+		perror("clock_getres");
+		return -1;
+	}
+
+	verbose("random=%s, ", yesno(is_random == CRYPTO_USE_RANDOM));
+	verbose("inner loops=%u, loops=%u, warm-up=%u s\n", l, n, warmup);
+	vverbose("Clock resolution is %jd ns\n", (intmax_t)ts.tv_sec *
+		 1000000000 + ts.tv_nsec);
+
+	if (tee_type == ECDSA || tee_type == ECDH) {
+		curve_id = get_curve_id(width_bits);
+		if (curve_id == TEE_ERROR_BAD_PARAMETERS)
+			return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	open_ta();
+
+	alloc_buffers(size, verbosity);
+	if (is_random == CRYPTO_USE_ZEROS)
+		memset((uint8_t *)in_shm.buffer, 0, size);
+	else
+		read_random(in_shm.buffer, size);
+	if (mode == MODE_DECRYPT && tee_type == RSA)
+		((unsigned char *)(in_shm.buffer))[0] = 0x00;
+
+	if (tee_type == X25519)
+		width_bits = WIDTH_BITS_25519;
+
+	prepare_obj(width_bits, tee_type);
+
+	if (tee_type == DH) {
+		read_random(keygen_dh_p, BITS_TO_BYTES(width_bits));
+		read_random(keygen_dh_g, DH_G_SIZE);
+		keygen_dh_p[0] |= 0x2; /* make sure the p is full width */
+		/* make sure keygen_dh_p is odd */
+		keygen_dh_p[BITS_TO_BYTES(width_bits) - 1] |= 0x1;
+
+		xtest_add_attr(&param_count, params, TEE_ATTR_DH_PRIME,
+			       keygen_dh_p, BITS_TO_BYTES(width_bits));
+		xtest_add_attr(&param_count, params, TEE_ATTR_DH_BASE,
+			       keygen_dh_g, DH_G_SIZE);
+	} else if (tee_type == ECDSA || tee_type == ECDH) {
+		xtest_add_attr_value(&param_count, params, TEE_ATTR_ECC_CURVE,
+				     curve_id, 0);
+	}
+
+	res = pack_attrs(params, param_count, &buf, &blen);
+	CHECK(res, "pack_attrs", return res;);

```

Comment:
s/`return res;`/`return -1;`/ for consistency in return codes.
Ditto at line 458.